File: desktop/electron/agent/tools.ts
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { runDesktopTool, type DesktopToolContext } from "../executor/runTool";
import { buildOpsLangChainTools, buildReadOnlyOpsTools, type OpsToolContext } from "./opsTools";
import {
  buildReadOnlyTopologyTools,
  buildTopologyLangChainTools,
  type TopologyToolContext,
} from "./topologyTools";
import {
  buildReadOnlyWorkspaceTools,
  buildWorkspaceLangChainTools,
  type WorkspaceToolContext,
} from "./workspaceTools";

export type AgentToolContext = DesktopToolContext & {
  resourceServerUrl?: string | null;
  workflowId?: string | null;
  stepId?: string | null;
};

function toTopologyContext(ctx: AgentToolContext): TopologyToolContext {
  return {
    workspaceRoot: ctx.workspaceRoot,
    resourceServerUrl: ctx.resourceServerUrl,
  };
}

function toOpsContext(ctx: AgentToolContext): OpsToolContext {
  return { workspaceRoot: ctx.workspaceRoot };
}

function toWorkspaceContext(ctx: AgentToolContext): WorkspaceToolContext {
  return {
    workspaceRoot: ctx.workspaceRoot,
    workflowId: ctx.workflowId,
    stepId: ctx.stepId,
  };
}

export function buildDesktopLangChainTools(ctx: AgentToolContext) {
  return [
    ...buildBaseDesktopTools(ctx),
    ...buildWorkspaceLangChainTools(toWorkspaceContext(ctx)),
    ...buildOpsLangChainTools(toOpsContext(ctx)),
    ...buildTopologyLangChainTools(toTopologyContext(ctx)),
  ];
}

export function buildReadOnlyDesktopTools(ctx: AgentToolContext) {
  return [
    ...buildBaseDesktopTools(ctx).filter((t) =>
      ["read_file", "list_dir", "git_status", "git_diff"].includes(t.name),
    ),
    ...buildReadOnlyWorkspaceTools(toWorkspaceContext(ctx)),
    ...buildReadOnlyOpsTools(toOpsContext(ctx)),
    ...buildReadOnlyTopologyTools(toTopologyContext(ctx)),
  ];
}

function buildBaseDesktopTools(ctx: DesktopToolContext) {
  return [
    tool(
      async ({ path }) => runDesktopTool(ctx, "read_file", { path }),
      {
        name: "read_file",
        description: "Read a UTF-8 text file relative to the workspace root.",
        schema: z.object({ path: z.string() }),
      },
    ),
    tool(
      async ({ path }) => runDesktopTool(ctx, "list_dir", { path }),
      {
        name: "list_dir",
        description: "List files and directories under a workspace-relative path.",
        schema: z.object({ path: z.string().default(".") }),
      },
    ),
    tool(
      async () => runDesktopTool(ctx, "git_status", {}),
      {
        name: "git_status",
        description: "Show git status for the workspace repository.",
        schema: z.object({}),
      },
    ),
    tool(
      async ({ path }) => runDesktopTool(ctx, "git_diff", { path }),
      {
        name: "git_diff",
        description: "Show git diff for the whole repo or a specific path.",
        schema: z.object({ path: z.string().default("") }),
      },
    ),
    tool(
      async ({ command }) => runDesktopTool(ctx, "run_shell", { command }),
      {
        name: "run_shell",
        description: "Run a shell command in the workspace directory.",
        schema: z.object({ command: z.string() }),
      },
    ),
  ];
}

File: backend/app/agent/tools/desktop/tools.py
from langchain_core.tools import tool

from app.agent.tools.desktop.executor_client import ExecutorError, call_executor


@tool
def read_file(path: str) -> str:
    """Read a UTF-8 text file relative to the workspace root."""
    try:
        return call_executor("read_file", {"path": path})
    except ExecutorError as exc:
        return f"error: {exc}"


@tool
def list_dir(path: str = ".") -> str:
    """List files and directories under a workspace-relative path."""
    try:
        return call_executor("list_dir", {"path": path})
    except ExecutorError as exc:
        return f"error: {exc}"


@tool
def git_status() -> str:
    """Show git status for the workspace repository."""
    try:
        return call_executor("git_status", {})
    except ExecutorError as exc:
        return f"error: {exc}"


@tool
def git_diff(path: str = "") -> str:
    """Show git diff for the whole repo or a specific path."""
    try:
        return call_executor("git_diff", {"path": path})
    except ExecutorError as exc:
        return f"error: {exc}"


@tool
def run_shell(command: str) -> str:
    """Run a shell command in the workspace directory. Use for build/test commands."""
    try:
        return call_executor("run_shell", {"command": command})
    except ExecutorError as exc:
        return f"error: {exc}"


def build_desktop_tools() -> list:
    return [read_file, list_dir, git_status, git_diff, run_shell]
